content/client-lib-development-guide/features.textile

```diff
@@ -103,10 +103,12 @@ h3(#rest-auth). Auth
 ** @(RSA4a)@ When a @token@ or @tokenDetails@ is used to instance the library, and no means to renew the token is provided (either an API key, @authCallback@ or @authUrl@), if the server responds with a token error (401 HTTP status code and an Ably error value @40140 <= code < 40150@), then the client library should indicate an error, not retry the request and in the case of the realtime library, transition the connection to the @FAILED@ state
 ** @(RSA4b)@ When the client does have a means to renew the token automatically, and the token has expired or the server has responded with a token error (@statusCode@ value of 401 and error @code@ value in the range @40140 <= code < 40150@), then the client should automatically make a single attempt to reissue the token and resend the request using the new token. If the token creation failed or the subsequent request with the new token failed due to a token error, then the request should result in an error
-** @(RSA4c)@ If an attempt by the realtime client library to authenticate is made using the @authUrl@ or @authCallback@, and the request to @authUrl@ fails, the callback @authCallback@ results in an error, the provided token is in an invalid format, or the attempt times out after "@realtimeRequestTimeout@":#DF1b, then:
+** @(RSA4c)@ If an attempt by the realtime client library to authenticate is made using the @authUrl@ or @authCallback@, and the request to @authUrl@ fails (unless @RSA4d@ applies), the @authCallback@ results in some error (unless @RSA4d@ or @RSA4e@ apply), the provided token is in an invalid format, or the attempt times out after "@realtimeRequestTimeout@":#DF1b, then:
 *** @(RSA4c1)@An @ErrorInfo@ with @code@ @80019@ and description of the underlying failure should be emitted with the state change, in the @errorReason@ and/or in the callback as appropriate
 *** @(RSA4c2)@If the connection is @CONNECTING@, then the connection attempt should be treated as unsuccessful, and as such the connection should transition to the @DISCONNECTED@ or @SUSPENDED@ state as defined in "RTN14":#RTN14 and "RTN15":#RTN15
 *** @(RSA4c3)@If the connection is @CONNECTED@, then the connection should remain @CONNECTED@
+** @(RSA4d)@ If a request by a realtime client to an @authUrl@ results in an HTTP 403 response, then the client library should transition to the @FAILED@ state, with the connection @errorReason@ being set to an @ErrorInfo@ with @code@ @40300@ and some appropriate message.
+** @(RSA4e)@ If an @authCallback@ attempt or @authUrl@ request by a realtime client results in an Ably @ErrorInfo@, the client library should transition to the @FAILED@ state, with the connection @errorReason@ being set to the error. (This applies only where the response received over the wire is actually an @ErrorInfo@; any errors generated by the client, such as network errors, must be handled per @RSA4c@).
```

Comment:
I am really not sure how the library is going to know it's an Ably @ErrorInfo@ specifically, especially with typed languages where the IDL makes no mention of an ErrorInfo object `authCallback: (String | TokenDetails | TokenRequest) →? // RSC14c, RSA4, TO3j5, AO2b`.

I think this is quite unclear as to why this spec item exists, what the purpose is, and how it would be used.  

I recommend that RSA4d covers only authUrl and we explain that unless an unauthorized response is received from an auth URL, all other failures are not considered terminal and thus the client library moves to the disconnected state so that an auth attempt will be retried.

For RSA4e, we should make it clear that an authCallback failure in most cases would be considered not terminal and therefore the library will move to the disconnected state (in the case of realtime), and reattempt a connection at a later time thus triggering another authCallback.  If we do want to support passing an error to the callback, then we should consider adding a new argument to the callback IMHO which is error, which will be null when successful, and contain an Ably ErrorInfo when unsuccessful.  If the error status code is 403, then it too will be treated as terminal and thus move the connection to the FAILED state